sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -2049,6 +2049,17 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val NESTED_PREDICATE_PUSHDOWN_ENABLED =
+    buildConf("spark.sql.optimizer.nestedPredicatePushdown.enabled")
+      .internal()
+      .doc("When true, Spark tries to push down predicates for nested columns and or names " +
+        "containing `dots` to data sources. Currently, Parquet implements both optimizations " +
+        "while ORC only supports predicates for names containing `dots`. The other data sources" +
+        "don't support this feature yet.")
+      .version("3.0.0")
+      .booleanConf
+      .createWithDefault(true)
```

Comment:
Will we port this back to branch-3.0?

Reply:
Besides of https://github.com/apache/spark/pull/27728#discussion_r397742247, one more concern about enabling by default is, after this is enabled, we will push down `a.b` as `b` in `a` by default.

I don't think DSv1 pushed down non-existent columns before; however, now DSv1 implementations should understand non-existent column  named `a.b`.

I think we shouldn't assume the DSv1 downstream sources handle non-existent column handling by default. Think about constructing query strings from filters like JDBC - it will fail and the implementations have to be fixed, or this configuration has to be disabled. However, this configuration is none or all.

Reply:
Making a configuration that takes a list of sources could be done separately in a separate PR to make this PR smaller.

Reply:
Since the filter apis will be enhanced to support nested columns and column name containing `dots`, it will be nice to introduce it in a major release. 

It's a good idea! We can make another PR to turn this feature on for specific data sources in a separate PR. This PR already grows too big.

Thanks!

Reply:
FYI @HeartSaVioR, are you interested in this followup out of curiosity?

Reply:
Thanks for pinging me!

I might not be able to do this very soon, so please go ahead if someone is also interested or planning to. If that would be OK I deal with this in a week (or maybe even a couple of weeks), yeah I'm interested.

Reply:
Sure, I think it should be fine.